dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorInlineAsm.java

```diff
@@ -0,0 +1,550 @@
+    private final HashMap<String, Label> labelsDefined = new HashMap<>();
+    // used to keep track of which asm register should map to the llvm return register if it is an aggregate type
+    private final List<Register> pendingRegisters = new ArrayList<>();
+    // holds the LLVM registers that are passed (as args) to the the asm side
+    private final List<Expression> argsRegisters;
+    // expected type of RHS of a comparison.
+    private Type expectedType;
+    // map from RegisterID to the corresponding asm register
+    private final HashMap<Integer, Register> asmRegisters = new HashMap<>();
+
+    public VisitorInlineAsm(Function llvmFunction, Register returnRegister, List<Expression> llvmArguments) {
+        this.llvmFunction = llvmFunction;
+        this.returnRegister = returnRegister;
+        this.argsRegisters = llvmArguments;
+    }
+
+    // Returns the size of the return register
+    // null / void -> 0
+    // i32 / bool -> 1
+    // aggregateType -> the size of the aggregate
+    private int getSizeOfReturnRegister() {
+        if (this.returnRegister == null) {
+            return 0;
+        }
+        Type returnType = this.returnRegister.getType();
+        if (returnType instanceof IntegerType || returnType instanceof BooleanType) {
+            return 1;
+        } else if (returnType instanceof AggregateType at) {
+            return at.getTypeOffsets().size();
+        } else if (returnType instanceof VoidType) {
+            return 0;
+        } else {
+            throw new ParsingException("Unknown inline asm return type " + returnType);
+        }
+    }
+
+    // Tells if the returnRegister is an AggregateType
+    private boolean isReturnRegisterAggregate() {
+        return getSizeOfReturnRegister() > 1;
+    }
+
+    // Tells if the registerID is mapped to the returnRegister
+    private boolean isPartOfReturnRegister(int registerID) {
+        return registerID < getSizeOfReturnRegister();
+    }
+
+    // Given a string of a label, it either creates a new label, or returns the existing one if it was already defined
+    private Label getOrNewLabel(String labelName) {
+        if (!this.labelsDefined.containsKey(labelName)) {
+            this.labelsDefined.put(labelName, EventFactory.newLabel(labelName));
+        }
+        return this.labelsDefined.get(labelName);
+    }
+
+    // This function lets us know which type we need to assign to the created asm register.
+    // In order to do so, we have to understand which llvm register it is going to be mapped to.
+    // Given the registerID of the register e.g. $2 -> registerID = 2
+    // returns the type of the llvm register it is mapped to by the clobbers
+    // if it is referencing the return register, return its type.
+    private Type getLlvmRegisterTypeGivenAsmRegisterID(int registerID) {
+        Type registerType;
+        if (isPartOfReturnRegister(registerID)) {
+            if (returnRegister.getType() instanceof AggregateType at) {
```

Comment:
You have a `isReturnRegisterAggregate` function. Also, you should be consistent how you check for aggregates: either you use `instanceof` or you check the size, but not both if possible